docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+
+Although we could put the responsibility for checking and updating disruption budgets
+solely on the client, it is safer and more convenient if we implement that functionality
+in the API server. Thus we will introduce a new `/evict` subresource on pod. It is similar to
+today's "delete" on pod except
+  * It will be rejected if the deletion would violate disruption budget. (See how
+  Deployment handles failure of /rollback for ideas on how clients could handle failure
+  of `/evict`.) There are two possible ways to implement this:
+    * For the initial implementation, this will be accomplished by the API server just
+  looking at the `DisruptionBudgetStatus` and seeing if the disruption would violate the
+  `DisruptionBudgetSpec`. In this approach, we assume a disruption budget controller
+  keeps the `DisruptionBudgetStatus` up-to-date by observing all pod deletions and
+  creations in the cluster, so that an approved disruption is quickly reflected in the
+  `DisruptionBudgetStatus`. Of course this approach does allow a race in which one or
+  more additional disruptions could be approved before the first one is reflected in the
+  `DisruptionBudgetStatus`.
+    * Thus a subsequent implementation will have the API server explicitly debit the
+  `DisruptionBudgetStatus` when it accepts an `/evict`. (There still needs to be a
+  controller, to keep the shard strength status up-to-date when replacement pods are
+  created after an eviction; the controller may also be necessary for the rate status
+  depending on how rate is represented, e.g. adding tokens to a bucket at a fixed rate.)
+  Once etcd support multi-object transactions (etcd v3), the debit and pod deletion will
+  be placed in the same transaction.
+    * Note: For the purposes of disruption budget, a pod is considered to be disrupted as soon as its
+  graceful termination period starts (so when we say "delete" here we do not mean
+  "deleted from etcd" but rather "graceful termination period has started").
+  * It will allow clients to communicate additional parameters when they wish to delete a
+  pod. (In the absence of the `/evict` subresource, we would have to create a pod-specific
+  type analogous to `api.DeleteOptions`.)
+
+We will make `kubectl delete pod` use `/evict` by default, and require a command-line
+flag to delete the pod directly.
+
+We will add to `NodeStatus` a bounded-sized list of signatures of pods that should avoid
+that node (provisionally called `PreferAvoidPods`). One of the pieces of information
+specified in the `/evict` subresource is whether the eviction should add the evicted
+pod's signature to the corresponding node's `PreferAvoidPods`. Initially the pod
+signature will be a
+[controllerRef](https://github.com/kubernetes/kubernetes/issues/14961#issuecomment-183431648),
+i.e. a reference to the pod's controller. Controllers are responsible for garbage
+collecting, after some period of time, `PreferAvoidPods` that point to them, but the API
```

Comment:
Yeah, sorry for the confusing terminology. As @HaiyangDING inferred, I just meant the controller should remove itself from the list after some period of time.